[13.0][ADD] product_variant_change_attribute_value

sebalix opened the pull request:
Same PR than https://github.com/OCA/product-variant/pull/215 but based on an older commit (ping @TDu).

With standard Odoo there is no way to change the attribute values assigned
to a product variant.

This module adds a wizard that can be access through the product variant
tree view context action menu.

The wizard allows to change or remove attribute values on all product
variant selected.

When an attribute value is removed from all variant of a template,
it will also be removed from the configuration of the related product template.

Ref. 2090

TDu committed:
[13.0][ADD] product_variant_change_attribute_value

With standard Odoo there is no way to change the attribute values assigned
to a product variant.

This module adds a wizard that can be access through the product variant
tree view context action menu.

The wizard allows to change or remove attribute values on all product
variant selected.

When an attribute value is removed from all variant of a template,
it will also be removed from the configuration of the related product template.

simahawk commented on product_variant_change_attribute_value/wizards/product_variant_attribute_value_action.py:
```diff
@@ -0,0 +1,35 @@
+    @api.depends("attribute_action")
+    def _compute_selectable_attribute_value_ids(self):
+        for rec in self:
+            attribute_ids = self.attribute_id.value_ids.ids
```
1. there's nothing depending on `attribute_action`
2. I think we can drop `compute` and make this field related to `attribute_id`

simahawk commented on product_variant_change_attribute_value/wizards/product_variant_attribute_value_wizard.py:
```diff
@@ -0,0 +1,135 @@
+    _description = "Wizard to change attriubtes on product variants"
+
+    product_ids = fields.Many2many(
+        "product.product", default=lambda self: self._default_product_id()
```
```suggestion
        "product.product", default=lambda self: self._default_product_ids()
```

simahawk commented on product_variant_change_attribute_value/wizards/product_variant_attribute_value_wizard.py:
```diff
@@ -0,0 +1,135 @@
+        return self.env["product.product"].browse(self._context.get("default_res_ids"))
+
+    def _default_attributes_action_ids(self):
+        p = self.env["product.product"].browse(self._context.get("default_res_ids"))
```
```suggestion
        p = self._default_product_ids()
```

simahawk commented on product_variant_change_attribute_value/wizards/product_variant_attribute_value_wizard.py:
```diff
@@ -0,0 +1,135 @@
+            self.update_variant_value(product)
+
+    def _is_attribute_value_being_used(self, variant_id, attribute_value):
+        """Check if attribute value is still in used on any variant of a template."""
```
```suggestion
        """Check if attribute value is still used by a variant."""
```

simahawk commented on product_variant_change_attribute_value/wizards/product_variant_attribute_value_action.py:
```diff
@@ -0,0 +1,35 @@
+    selectable_attribute_value_ids = fields.Many2many(
+        "product.attribute.value", compute="_compute_selectable_attribute_value_ids"
+    )
+    replaced_by = fields.Many2one(
```
```suggestion
    replaced_by_id = fields.Many2one(
```

simahawk commented on product_variant_change_attribute_value/wizards/product_variant_attribute_value_wizard.py:
```diff
@@ -0,0 +1,135 @@
+                continue
+            elif action == "delete":
+                pass
+            elif action == "replace" and pav_replacement:
```
split this to `_replace_attribute`

simahawk commented on product_variant_change_attribute_value/wizards/product_variant_attribute_value_wizard.py:
```diff
@@ -0,0 +1,135 @@
+                    [
+                        ("attribute_line_id", "=", tpl_attr_line.id),
+                        ("product_attribute_value_id", "=", pav_replacement.id),
+                    ]
```
```suggestion
                    ], limit=1
```

simahawk commented on product_variant_change_attribute_value/wizards/product_variant_attribute_value_wizard.py:
```diff
@@ -0,0 +1,135 @@
+                    [
+                        ("attribute_line_id", "=", tpl_attr_line.id),
+                        ("product_attribute_value_id", "=", pav.id),
+                    ]
```
```suggestion
                    ], limit=1
```

simahawk commented on product_variant_change_attribute_value/wizards/product_variant_attribute_value_wizard.xml:
```diff
@@ -0,0 +1,84 @@
+        <field name="arch" type="xml">
+            <form string="Product Variant Change Attribute">
+                <sheet>
+                    <!-- <div>Updating 8 variant(s) selected out of 2 product template.</div> -->
```
just a TODO?

simahawk commented on product_variant_change_attribute_value/tests/test_product_variant_change_attribute_value.py:
```diff
@@ -0,0 +1,124 @@
+
+        cls.wizard = cls.env["variant.attribute.value.wizard"]
+
+    def change_action(self, value, action, replaced_by=False):
```
these methods should be private

simahawk commented on product_variant_change_attribute_value/tests/test_product_variant_change_attribute_value.py:
```diff
@@ -0,0 +1,124 @@
+            [
+                ("attribute_line_id", "=", attribute_line.id),
+                ("product_attribute_value_id", "=", attribute_value.id),
+            ]
```
```suggestion
            ], limit=1
```

simahawk commented on product_variant_change_attribute_value/tests/test_product_variant_change_attribute_value.py:
```diff
@@ -0,0 +1,124 @@
+        """Check removing an attribute value on ALL variants of a template."""
+        self.assertTrue(self.is_value_on_variant(self.variant_1, self.steel))
+
+        self.wiz = self.wizard.with_context(default_res_ids=self.variants).create({})
```
no need to assign it to `self`, pass `wiz` around where needed.

simahawk commented on product_variant_change_attribute_value/readme/DESCRIPTION.rst:
```diff
@@ -0,0 +1,11 @@
+With standard Odoo there is no way to change the attribute values assigned
+to a product variant.
+
+This module adds a wizard that can be access through the product variant
+tree view context action menu.
+
+The wizard allows to change or remove attribute values on all product
+variant selected.
+
+When an attribute value is removed from all variant of a template,
+it will also be removed from the configuration of the related product template.
```
```suggestion
In standard Odoo there is no way to change the attribute values assigned
to a product variant.

This module adds a wizard to change or remove attribute values on all product
variant selected.

When an attribute value is removed from all variants of a template,
it will also be removed from the configuration of the template.
```

Move "available through the product variant tree view context action menu." to `USAGE.rst`

TDu committed:
fixup! [13.0][ADD] product_variant_change_attribute_value

TDu committed:
fixup! fixup! [13.0][ADD] product_variant_change_attribute_value

simahawk commented on product_variant_change_attribute_value/wizards/product_variant_attribute_value_action.py:
```diff
@@ -8,9 +8,16 @@ class ProductVariantAttributeValueAction(models.TransientModel):
     product_attribute_value_id = fields.Many2one("product.attribute.value",)
     attribute_action = fields.Selection(
-        [("delete", "Delete"), ("replace", "Replace"), ("do_nothing", "Do Nothing")],
+        selection=lambda self: self._get_attibute_action_list(),
```
`selection="_selection_attribute_action"`

TDu replied:
I changed it

simahawk commented on product_variant_change_attribute_value/wizards/product_variant_attribute_value_wizard.py:
```diff
@@ -0,0 +1,144 @@
+    )
+
+    def _default_product_ids(self):
+        return self.env["product.product"].browse(self._context.get("default_res_ids"))
```
`default_res_ids` seems a custom key. Instead, simply set `default_product_ids` in the ctx and will work automatically w/out this method. All in all, it's just a default :wink:

TDu committed:
fixup! fixup! fixup! [13.0][ADD] product_variant_change_attribute_value

TDu committed:
fixup! fixup! fixup! fixup! [13.0][ADD] product_variant_change_attribute_value

TDu committed:
Add some tests for specific case

There may be an issue with the wizard with some attribute value not
being active when they are actually used by some variant.

But could not reproduce the error with the 2 new tests

TDu committed:
fixup! Add some tests for specific case

TDu committed:
fixup! fixup! Add some tests for specific case

simahawk commented:
closed in favour of https://github.com/OCA/product-variant/pull/222
